ENH: signal.lp2{lp,hp,bp,bs}: add array API standard support

ev-br opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->

towards https://github.com/scipy/scipy/issues/20678

#### What does this implement/fix?
<!--Please explain your changes.-->

Make `lp2{lp,hp,bp,bs}` and their `_zpk` cousins array API compatible.

#### Additional information
<!--Any additional information you think is important.-->

Got to have a bit of fun with probably least used parts of numpy: `np.trim_zeros`, anyone? Also `np.append` and my personal favorite, `np.resize`.

That said, changes are mostly mechanical.

ev-br committed:
MAINT: signal: convert lp2{lp,hp, bp, bs} to be array api compatible

ev-br committed:
MAINT: signal: convert lp2{lp,bp,hp}_zpk to be array api compatible

ev-br committed:
MAINT: appease the linter

ev-br committed:
MAINT: signal: bilinear_zpk array API

tylerjereddy reviewed (commented):
Just a minor contribution, but I found that marking three tests as eager-only was sufficient to make all backends (including JAX, CuPy) happy locally.

<details>

```diff
--- a/scipy/signal/tests/test_filter_design.py
+++ b/scipy/signal/tests/test_filter_design.py
@@ -1367,6 +1367,7 @@ class TestLp2lp:
 
 class TestLp2hp:
 
+    @skip_xp_backends(eager_only=True, reason="in-place item assignment")
     def test_basic(self, xp):
         b = xp.asarray([0.25059432325190018])
         a = xp.asarray(
@@ -1381,6 +1382,7 @@ class TestLp2hp:
 
 class TestLp2bp:
 
+    @skip_xp_backends(eager_only=True, reason="in-place item assignment")
     def test_basic(self, xp):
         b = xp.asarray([1])
         a = xp.asarray([1, 2, 2, 1])
@@ -1395,6 +1397,7 @@ class TestLp2bp:
 
 class TestLp2bs:
 
+    @skip_xp_backends(eager_only=True, reason="in-place item assignment")
     def test_basic(self, xp):
         b = xp.asarray([1])
         a = xp.asarray([1, 1])
```

</details>

I took a quick stab at using `xpx.at(...).set(...)` for one the source code paths, but it turned into a bit of a rabbit hole.

j-bowhay reviewed (commented):
Looks good, just one query

ev-br committed:
BUG: signal: make normalize accept object arrays, fix _align_num

Object arrays are explicitly used in
test_ltisys.py::TestSS2TF::test_simo_round_trip

ev-br committed:
TST: signal: temp workarounds for tf2sos being numpy-only

ev-br committed:
MAINT/TST: signal: avoid nested asarrays in convolve2d tests

This to be forward compatible with https://github.com/data-apis/array-api/pull/917
(planned for Array API spec 2025.12)

tylerjereddy committed:
TST: signal: add skips for non-eager backends

ev-br commented on scipy/signal/_filter_design.py:
```diff
@@ -1953,27 +2005,34 @@ def lp2hp(b, a, wo=1.0):
-        outa = a[::-1] * pwo
-        outb = resize(b, (d,))
+        outa = xp.flip(a) * pwo
+        outb = xp.concat((xp.zeros(n, dtype=b.dtype), ))
```
A clear error, thanks!

I keep mixing the behavior of `ndarray.resize` and `np.resize`, the former can be replicated with a concat, while here we need the latter.
Fixed.

ev-br committed:
TST: signal: add skips for dask

The problem seems to be with Dask's fancy indexing:

(Pdb) p arr[idxs]
*** numpy.exceptions.ComplexWarning: Casting complex values to real discards the imaginary part
(Pdb) p arr.compute()[idxs.compute()]
array([-3.53553391-3.53553391j, -3.53553391+3.53553391j])
(Pdb) p arr
dask.array<mul, shape=(2,), dtype=complex128, chunksize=(2,), chunktype=numpy.ndarray>
(Pdb) p idxs
dask.array<argsort, shape=(2,), dtype=int64, chunksize=(2,), chunktype=numpy.ndarray>
(Pdb) p arr.compute()
array([-3.53553391+3.53553391j, -3.53553391-3.53553391j])
(Pdb) p idxs.compute()
array([1, 0])

ev-br committed:
MAINT: signal: fix the next slew of lp2* related test failures

ev-br committed:
MAINT: signal: mention the dask issue in skips

ev-br commented:
Okay, all green finally, modulo unrelated jitter in the Windows jobs.

Getting the CI to green needed a couple of fixups:
- object arrays are still tested for, from `test_ltisys`
- a couple of temporary fixes for `tf2**` functions not converted yet,
- skips for https://github.com/dask/dask/issues/11883 and https://github.com/data-apis/array-api-compat/issues/292
- Tyler's patch for lazy backends (Thanks Tyler!)

j-bowhay commented on scipy/signal/_filter_design.py:
```diff
@@ -1778,30 +1801,37 @@ def normalize(b, a):
-        if np.allclose(col, 0, atol=1e-14):
+    for j in range(num.shape[-1]):
+        col = num[:, j]
+        if xp.all(xp.abs(col) <= 1e-14):
```
Is this equivalent since `rtol` has a default value?

ev-br replied:
In general it isn't indeed. Here specifically, the second argument of `np.allclose`, `b`, is zero. And `np.allclose` is defined for `rtol` relative to `b`:

```
absolute(a - b) <= (atol + rtol * absolute(b))
```
https://numpy.org/doc/stable/reference/generated/numpy.allclose.html

j-bowhay reviewed (approved):
Thanks, did another pass and everything seems to be in order!
